### DISCUSSION TITLE
<Link> prefetching behavior

### DISCUSSION THREAD

**Discussion by yousefelgoharyx** [vercel/next.js#60581] [Ideas]
### Goals

1. Prefetching only when it's safe to do so

### Background

I'm working on a project using Next 14 and app router. We have a route for logging users out. This route just deletes some cookies then redirects the user to another page.

The route exists in the app directory and the folder `logout` only has `route.ts`  as shown down
![image](https://github.com/vercel/next.js/assets/30435916/3c05e50b-9fb2-40e9-a3ae-5fbb83c48361)

Now the only missing piece is having a <Link> component that goes to /logout as shown down
![image](https://github.com/vercel/next.js/assets/30435916/d2de2ba7-68d5-443a-a10a-401a8f60e1a0)
 
 The problem is that Next will prefetch /logout if we don't pass prefetch={false}. This is really unsafe and in our case it leads to the user being signed out the moment he logs in to their account.
 

### Proposal

I haven't really thought of anything to solve this DX experience but an ideal solution would be that Next only prefetches routes that actually serve some content
since pages in app router are not allowed to set or delete cookies, it's safe to prefetch only routes that actually serve some content

**Comment by icyJoseph** [comment]
What happens when the user clicks on the link? Maybe you should switch it to a regular `a` tag. Anyhow, I think a form that gets submitted to `/auth/logout`, when a link/button is clicked is probably better for this use case.

**Reply by yousefelgoharyx** [reply]
Hmmm I usually use forms to mutate data on the server. So using a form to logout the current user seems a bit odd to me I don't know if it's ok to do so

However, my point is that Next prefetches routes arbitrary and that could be unsafe in many cases such as my case

**Reply by icyJoseph** [reply]
Isn't changing the logged in state a mutation?

And sure, a form is iffy, but one could also make the case that GET requests ought to be idempotent, AKA a query should NOT turn into a command.

**Comment by augustl** [comment]
How does Next actually handle prefetching for "dynamic" routes like this? Does it just invoke the entire route, including calls to `cookie().delete(...)` etc, when there's no `loading.tsx`?

**Comment by sabeerbikba** [comment]
> ### Goals
> 1. Prefetching only when it's safe to do so
> 
> ### Non-Goals
> _No response_
> 
> ### Background
> I'm working on a project using Next 14 and app router. We have a route for logging users out. This route just deletes some cookies then redirects the user to another page.
> 
> The route exists in the app directory and the folder `logout` only has `route.ts` as shown down ![image](https://private-user-images.githubusercontent.com/30435916/296345476-3c05e50b-9fb2-40e9-a3ae-5fbb83c48361.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzOTI0MDcsIm5iZiI6MTcyMTM5MjEwNywicGF0aCI6Ii8zMDQzNTkxNi8yOTYzNDU0NzYtM2MwNWU1MGItOWZiMi00MGU5LWEzYWUtNWZiYjgzYzQ4MzYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE5VDEyMjgyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY0MTZjNmExYWNkOTVmNTRkZGNmYjA4ZWU4ZjliNGVhZjFmMGU2ZGFiMmFmMTg2YTYyZWJiOTI5ZjM5OTNkNzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ttIWxfBUNLFhDPGu-DhaLpsN27f6YSP8wLHTt-xqsKM)
> 
> Now the only missing piece is having a component that goes to /logout as shown down ![image](https://private-user-images.githubusercontent.com/30435916/296345861-d2de2ba7-68d5-443a-a10a-401a8f60e1a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzOTI0MDcsIm5iZiI6MTcyMTM5MjEwNywicGF0aCI6Ii8zMDQzNTkxNi8yOTYzNDU4NjEtZDJkZTJiYTctNjhkNS00NDNhLWExMGEtNDAxYThmNjBlMWEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE5VDEyMjgyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZiMDQyMjM2ZWYzNmZjYTg4MGRhODkxZmQ0ZGY1Y2M3ZjBmMzgzOGU2MDA1YTNiNDBmYTVmOWFhOWJmNzg2ZGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.iwQwq0tt2TxDRGoNZPDoCogq7CYLTq4oStBoM19hcBs)
> 
> The problem is that Next will prefetch /logout if we don't pass prefetch={false}. This is really unsafe and in our case it leads to the user being signed out the moment he logs in to their account.
> 
> ### Proposal
> I haven't really thought of anything to solve this DX experience but an ideal solution would be that Next only prefetches routes that actually serve some content since pages in app router are not allowed to set or delete cookies, it's safe to prefetch only routes that actually serve some content

Try using html anchor tag:

```JavaScript 
<a href="/auth/logout">{t("logout")}</a>
```

**Comment by CoderDeltaLAN** [comment]
**Context**
- Next.js App Router
- `<Link />` prefetch causes unintended requests (e.g. `/logout`)
- You want to control or disable prefetch safely

**Why**
`<Link />` prefetches when it enters the viewport (prod only). This is by design to speed up navigations. However, prefetching *should not* trigger side-effects. If a route performs effects on GET (like logging out), that’s the root problem: side-effects must live behind POST/route actions, not GET.

**Fix (minimum viable)**
1) **Disable prefetch per link** when needed:
   ```tsx
   <Link href="/logout" prefetch={false}>Logout</Link>
   ```
2) **Guard the route** against GET side-effects:
   - For logout, use a POST Route Handler or a Server Action, then redirect.
   ```ts
   // app/logout/route.ts
   import { NextResponse } from 'next/server';
   export async function POST() {
     // clear cookies/session here (server-only)
     return NextResponse.redirect(new URL('/', process.env.NEXT_PUBLIC_BASE_URL));
   }
   ```
   And call it from a client button:
   ```tsx
   'use client'
   async function onLogout() {
     await fetch('/logout', { method: 'POST' });
     window.location.href = '/';
   }
   ```
3) **Global prefetch control** (opt-out where appropriate):
   - Prefer disabling prefetch only on links with side-effects.
   - For lists with many links, consider virtualization or set `prefetch={false}` on those items to reduce background traffic.

**Notes**
- Docs: Link prefetching (production-only, viewport-based): https://nextjs.org/docs/app/api-reference/components/link
- Good practice: routes that *change state* should require POST; GET should be safe/idempotent.

If this helps, please **Mark as answer** so others can find it.

**Comment by AssafMeidan** [comment]
## Real-world war story: `<Link href="/logout">` prefetch silently broke auth for 10+ hours

We spent ~10 hours and **migrated hosting platforms** debugging this. We initially blamed Railway for "stripping Set-Cookie headers," migrated to Vercel, and still had the same problem. It wasn't the hosting. It was `<Link href="/logout">` prefetch clearing our auth cookie on every page load.

**This affects any Next.js deployment — Vercel, Railway, AWS, self-hosted, even localhost in production mode.**

---

### Setup

- Next.js 15, App Router
- Cookie-based JWT auth (`HttpOnly`, `Secure`, `SameSite=Lax`)
- Logout was a **GET** route handler that clears the session cookie
- Sidebar had `<Link href="/logout">` pointing to the logout route

### Symptoms

- ✅ Login succeeded — dashboard rendered, user info visible in sidebar
- ❌ Clicking **any** link → redirected back to login
- ✅ `curl` with the same cookie worked perfectly (200 on all protected routes)
- ✅ Cookie storage confirmed working via isolated tests (both 303 redirect and 200+meta-refresh)
- ❌ Both Chrome and Firefox failed identically
- ❌ Incognito mode failed too

### Root cause

Next.js `<Link>` **prefetches its target on page load**. Our sidebar had `<Link href="/logout">`, and the `/logout` GET route handler called `clearSessionCookie()`. So every time the dashboard loaded, Next.js prefetched `/logout`, silently wiping the auth cookie before the user clicked anything.

The cookie was being set correctly. Something was **deleting** it immediately after.

### Why it took so long to find

1. **We asked the wrong question for 9+ hours.** We kept asking "why isn't the cookie being stored?" The right question was "**what's clearing it?**"
2. **`curl` worked** — because `curl` doesn't prefetch links. We drew the wrong conclusion from this ("must be a browser/hosting issue").
3. **We blamed the hosting platform.** Railway was "stripping Set-Cookie headers." We migrated to Vercel. Same bug. It was never the hosting.
4. **We rewrote the login flow 3 times** — tried `fetch()` + JSON, form POST + 303 redirect, raw HTML + meta-refresh, static HTML outside React entirely. None helped because the login was never the problem.
5. **We didn't open DevTools Network tab early enough.** The prefetch GET to `/logout` was visible in plain sight the entire time.
6. A simple `grep clearSessionCookie` would have traced the full code path in 5 minutes.

### The fix (not just `prefetch={false}`)

**Logout should be POST-only.** GET requests must be idempotent — they should never have side effects like clearing cookies.

```typescript
// app/logout/route.ts — POST only, never GET
export async function POST(request: Request) {
  const res = NextResponse.redirect(new URL('/login', request.url), 303);
  clearSessionCookie(res);
  return res;
}
```

```tsx
// Sidebar — <form> instead of <Link>
<form action="/logout" method="POST">
  <button type="submit" className="...">
    <LogOut className="h-3 w-3" />
    Logout
  </button>
</form>
```

`prefetch={false}` is a band-aid. Making logout POST-only is the proper fix because:

- **Prefetch only sends GET requests** — POST is never prefetched
- **POST can't be triggered accidentally** by the browser, crawlers, or extensions
- **It follows REST semantics** — logout is a state-changing operation and should never be GET

### Bonus discovery: React hydration breaks `Set-Cookie` on form POST

While debugging, we also discovered a separate (related) issue: **React hydration intercepts `<form method="POST">` submissions from Next.js pages** — even Server Components get hydrated, converting native form POSTs into `fetch()` calls. Browsers do **not** store `Set-Cookie` headers from `fetch()` responses (per the Fetch spec).

This means if your login form is a Next.js page component, the browser will never store the session cookie from the POST response — even though the server sends it correctly.

**Workaround:** Serve the login form as raw HTML from a route handler (`GET /api/auth/login`), not a Next.js page. Zero React/JS = native form POST = browser stores the cookie.

### Suggestions for the Next.js team

1. **Add a dev-mode warning** when `<Link>` points to a route handler that modifies cookies or performs side effects on GET
2. **Document in the Auth guide** that logout must be POST, not GET
3. **Consider skipping prefetch** for route handlers that call `cookies().delete()` or set `Set-Cookie` headers with `Max-Age=0`

### TL;DR

If your Next.js app loses auth cookies immediately after login — check if you have a `<Link>` pointing to a logout (or any cookie-clearing) route. The prefetch fires a GET request on page load, silently wiping your session before the user does anything. Make logout POST-only.
